include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -8105,7 +8115,10 @@ FEEvaluation<dim,
              n_q_points_1d,
              n_components_,
              Number,
-             VectorizedArrayType>::reinit(const unsigned int cell_index)
+             VectorizedArrayType>::
+  reinit(
+    const unsigned int                                             cell_index,
+    internal::MatrixFreeFunctions::ShapeInfo<VectorizedArrayType> *shape_info)
```

Comment:
I am not sure if it is wise to provide users the possibility to pass here a user shape_info. `FEEvaluation` might become easily inconsistent, despite the two new asserts. Question: what is stopping you to use `FEEvaluationImpl` directly in your user code - at least for now?

Reply:
Of course this can also be done in user code - my thought was the option could also be useful for someone else. What or where are the inconsistencies you fear?

Reply:
> What or where are the inconsistencies you fear?

`shape_info` contains the shape functions evaluated at the quadrature points and `mapping_info` the mapping info at the same points. MatrixFree guarantees internally that these are consistent. However, if you pass in a user-defined `shape_info` this is not guaranteed anymore: you could have the same number of quadrature points but at different positions. In your particular case, the points are simply shifted by a fixed interval (which seems to be consistent). But there are no way (?) to check if the data is consistent.

Personally, I am a bit restrained regarding the extension of `FEEvaluation`, since I am not sure how much benefit such a functionality would have. In your program, you are using this function but not in a vectorized fashion if I see it correctly.

Reply:
It is a bit of tradeoff where to put things. `FEEvaluationImpl` or the outer access function `internal::FEEvaluationFactory<dim, Number, VectorizedArrayType>::evaluate` (or `SelectEvaluator` if one wants the templated version) require a long chain of arguments to be passed in, in particular `FEEvaluationBase::{values,gradients,hessians}_quad`, or to do the quadrature point operations manually. That is not really nice. 
A follow-up problem to which there is no solution current is the fact that the `JxW` and `jacobians` entries do only match in case the mapping is affine and if the evaluation points are merely shifts of the unit cube, otherwise an arbitrary `shape_info` might lead to complete nonsense. That is unavoidable unless we also touch the mapping data.

Regarding the inconsistency: I see it mostly for the development use that we could miss to pass in the argument somewhere and hence get an inconsistent result. To ensure that we made no mistake, we would need  to think how to test all possibilities thoroughly. That would be somewhat challenging. On the other hand, from the user side I can see little that can go wrong, because a fresh reinit restores the initial `shape_info` (aka `this->data`), thereby making sure it does not stay around longer than necessary.

But I would say this is up to discussion. One thing is clear, we need to do more work before the release. At least there are hopes that we can get the `RaviartThomas` and `Nedelec` functionality into `FEEvaluation`, which would requiring touching at least all internal interfaces again.